crml/staking/src/rewards/mod.rs

```diff
@@ -151,17 +167,49 @@ where
 				Self::calculate_npos_payouts(&validator, *validator_commission, stake_map, total_payout_share)
 			})
 			.for_each(|(account, payout)| {
-				total_payout_imbalance.maybe_subsume(T::CurrencyToReward::deposit_into_existing(&account, payout).ok());
+				EraRemainedPayouts::<T>::mutate(|p| p.push((account, payout)));
 			});
 
-		// Any unallocated reward amount can go to the development fund
-		let remainder = total_payout.saturating_sub(total_payout_imbalance.peek());
-		total_payout_imbalance.subsume(
-			T::CurrencyToReward::deposit_into_existing(&T::TreasuryModuleId::get().into_account(), remainder)
-				.unwrap_or_else(|_| PositiveImbalanceOf::<T>::zero()),
-		);
+		EraRemainedRewardAmount::<T>::put(total_payout.saturating_sub(total_payout_imbalance.peek()));
+	}
+
+	fn process_reward_payouts(remained_blocks: Self::BlockNumber) -> Weight {
+		let remained_payouts = EraRemainedPayouts::<T>::get().len();
+		let quota = Self::calculate_payout_quota(remained_payouts, remained_blocks);
+		if quota == 0 {
+			return T::WeightInfo::process_zero_payouts();
+		}
+
+		let weight = T::WeightInfo::process_reward_payouts(quota);
+
+		EraRemainedPayouts::<T>::mutate(|p| {
+			let (a, m) = p.pop().unwrap_or_default();
+			let mut total_payout_imbalance =
+				T::CurrencyToReward::deposit_into_existing(&a, m).unwrap_or_else(|_| PositiveImbalanceOf::<T>::zero());
+
+			for _ in 1..quota {
+				if let Some((a, m)) = p.pop() {
+					total_payout_imbalance.maybe_subsume(T::CurrencyToReward::deposit_into_existing(&a, m).ok());
+				}
+			}
+
+			let remainder = EraRemainedRewardAmount::<T>::get().saturating_sub(total_payout_imbalance.peek());
 
-		Self::deposit_event(RawEvent::RewardPayout(total_payout_imbalance.peek(), remainder));
+			if p.len() > 0 {
+				EraRemainedRewardAmount::<T>::put(remainder);
+				return;
+			}
+			EraRemainedRewardAmount::<T>::put(BalanceOf::<T>::zero());
+
+			// Any unallocated reward amount can go to the development fund
+			total_payout_imbalance.subsume(
+				T::CurrencyToReward::deposit_into_existing(&T::TreasuryModuleId::get().into_account(), remainder)
+					.unwrap_or_else(|_| PositiveImbalanceOf::<T>::zero()),
+			);
+			Self::deposit_event(RawEvent::RewardPayout(total_payout_imbalance.peek(), remainder));
```

Comment:
we should log the event on line 200 when we early exit also

Reply:
Logging is now totally improved.